windows/src/global/delphi/chromium/Keyman.System.CEFManager.pas

```diff
@@ -141,14 +140,68 @@ procedure TCEFManager.RegisterWindow(cef: IKeymanCEFHost);
 
 function TCEFManager.Start: Boolean;
 begin
+  if GlobalCEFApp.ProcessType <> ptBrowser then
+  begin
+    LaunchWatchdog;
+  end;
+
   Result := GlobalCEFApp.StartMainProcess;
 end;
 
 function TCEFManager.StartSubProcess: Boolean;
 begin
+  LaunchWatchdog;
   Result := GlobalCEFApp.StartSubProcess;
 end;
 
+procedure TCEFManager.LaunchWatchdog;
+
+  function GetParentProcess: THandle;
+  var
+    Snapshot: THandle;
+    ProcessEntry: TProcessEntry32;
+    CurrentProcessId: DWORD;
+  begin
+    Snapshot := CreateToolhelp32Snapshot(TH32CS_SNAPPROCESS, 0);
+    if Snapshot = INVALID_HANDLE_VALUE then
+      Exit(0);
+
+    FillChar(ProcessEntry, sizeof(TProcessEntry32), 0);
+    ProcessEntry.dwSize := sizeof(TProcessEntry32);
+    CurrentProcessId := GetCurrentProcessId;
+
+    if Process32First(Snapshot, ProcessEntry) then
+    begin
+      repeat
+        if ProcessEntry.th32ProcessID = CurrentProcessId then
+          Break;
+      until not Process32Next(Snapshot, ProcessEntry);
+    end;
+
+    CloseHandle(Snapshot);
+
+    if ProcessEntry.th32ProcessID <> CurrentProcessId then
+      Exit(0);
+
+    Result := OpenProcess(SYNCHRONIZE, False, ProcessEntry.th32ParentProcessID);
+  end;
+
+var
+  hParentProcess: THandle;
+begin
+  hParentProcess := GetParentProcess;
+  if hParentProcess <> 0 then
+  begin
+    TThread.CreateAnonymousThread(
+      procedure
+      begin
+        WaitForSingleObject(hParentProcess, INFINITE);
+        ExitProcess(0);
```

Comment:
There may be an implication I'm missing due to comparative lack of familiarity with Delphi code, but does this just exit the child immediately once the parent process completes?  Was there supposed to be a `Cleanup` call here?

Reply:
This runs only in the context of the child process. It will normally only happen in an abnormal termination of the parent process (because otherwise the main thread of the child process receives an orderly request to shut down, and does so), but in the situation where the parent process disappears without making that request, this lets the child process do the work.

IMO, this is a flaw in Chromium Embedded Framework but AFAICT it has not been resolved and this was the suggested fix, e.g. at https://magpcss.org/ceforum/viewtopic.php?p=37820&sid=11f6fcabc4089e41283bd5b9ec17267e#p37820 which I have taken and cleaned up because it seemed appropriate.

Reply:
Could probably use a copy of that link as a comment in the code.